src/Deinterleave.cpp

```diff
@@ -198,202 +384,345 @@ class Deinterleaver : public IRGraphMutator {
     Expr visit(const VectorReduce *op) override {
         std::vector<int> input_lanes;
         int factor = op->value.type().lanes() / op->type.lanes();
-        for (int i = starting_lane; i < op->type.lanes(); i += lane_stride) {
+        input_lanes.reserve(factor * lanes.size());
+        for (int lane : lanes) {
             for (int j = 0; j < factor; j++) {
-                input_lanes.push_back(i * factor + j);
+                input_lanes.push_back(lane + j);
             }
         }
-        Expr in = Shuffle::make({op->value}, input_lanes);
-        return VectorReduce::make(op->op, in, new_lanes);
+        ScopedValue<DemandedLanes> old_lanes(lanes, input_lanes);
+        return VectorReduce::make(op->op, mutate(op->value), lanes.size());
     }
 
+    // This is essentially the same as `Ramp`/`Shuffle` handling,
+    // just specialized.
+    //
+    // FIXME: do we really prefer/want to retain the Broadcast
```

Comment:
The simplifier turns a shuffle of a broadcast into a smaller shuffle. Can't we just do the same here?

Reply:
> The simplifier turns a shuffle of a broadcast into a smaller broadcast. Can't we just do the same here?

Well, that was the question i was asking :)
I assume it's the same story for `Ramp`.
Thanks!